### DISCUSSION TITLE
`InvoiceReceivedPayment` or `InvoicePaymentSettled`?

### DISCUSSION THREAD

**Discussion by deadcoder0904** [btcpayserver/btcpayserver#5634] [Technical Support] [answered]
btc takes 5-10 minutes (sometimes 5, sometimes 7, sometimes 9) to come into the account.

so i am curious what do you guys do?

is `InvoiceReceivedPayment` okay to confirm that i got payment or i need `InvoicePaymentSettled`?

ik btc payments are final but can someone take it back after `InvoiceReceivedPayment` or can i just use that directly without confirming?

what is the common practice?

**Comment by dennisreimann** [comment, marked as answer]
Theoretically, payments which aren't confirmed (`InvoiceReceivedPayment`) can be double-spend, that's why you should use `InvoicePaymentSettled` to ensure the payment got the number of confirmations your store requires. This is configurable in the store settings.

**Reply by deadcoder0904** [reply]
yeah i saw the confirmations thingy... should i keep it as it is? its currently 1 (found under `Wallets > Bitcoin > Payment`)

![image](https://github.com/btcpayserver/btcpayserver/assets/16436270/d451000a-41f6-4b2e-a1ca-da30468e8e9a)

i'll use `InvoicePaymentSettled` & make customers wait for like 5-10 minutes before adding them to database.

**Reply by dennisreimann** [reply]
Yeah, keep it like that to be safe :)

**Reply by deadcoder0904** [reply]
@dennisreimann i used `InvoiceSettled` as recommended [here](https://github.com/btcpayserver/btcpayserver/discussions/5633#discussioncomment-8017947). lmk if its correct :)

**Reply by dennisreimann** [reply]
Yes, an invoice can have multiple (partial) payments, `InvoiceSettled` informs about the whole amount being paid.

**Reply by deadcoder0904** [reply]
@dennisreimann unfortunately, with `InvoiceSettled` i have a problem getting crypto payment & fees bcz the api isn't exposing it on `InvoiceSettled` but only on `InvoicePaymentSettled`? in such case, i'd have to add it up if partial payments. idk how that will work.

is there a way to get final amounts of crypto paid & fees any other way? or should i just drop that idea?

this i only get on `InvoicePaymentSettled`:

```json
{
	"afterExpiration": false,
	"paymentMethod": "BTC",
	"payment": {
			"id": "5f3bd1c8c83d360344d42fad16c2cb14ff69228203d3ce32025bb37053b512d7-0",
			"receivedDate": 1705050392,
			"value": "0.00080681",
			"fee": "0.0",
			"status": "Settled",
			"destination": "tb1q2lyl3dk7ctevus704y0lg85jg9l2vrkp0e93a5"
	},
	"deliveryId": "WzXpSwyZFR2UUpmcFJAeBy",
	"webhookId": "7vBiYUikVa8vyzdUN86E95",
	"originalDeliveryId": "WzXpSwyZFR2UUpmcFJAeBy",
	"isRedelivery": false,
	"type": "InvoicePaymentSettled",
	"timestamp": 1705051193,
	"storeId": "QDzp4UdVVyTuUm6k51uSz9ShtUCA6yz9jcCHZvuFomF",
	"invoiceId": "52sVVJiFLZ25ecKmZLQHTT",
	"metadata": {
			"orderId": "mbm3eq0cmbevlf4aj4o4dssn",
			"buyerEmail": "lol94@gmail.com",
			"project_name": "BTC Ebook"
	}
}
```

**Reply by dennisreimann** [reply]
You can use the [Greenfield API](https://docs.btcpayserver.org/API/Greenfield/v1/) to request the invoice and its individual sub-objects.

**Reply by deadcoder0904** [reply]
yes, i can definitely do that. but i do think a final crypto object should be exposed in `InvoiceSettled`. i checked & its nowhere exposed other than `InvoicePaymentSettled` & `InvoiceReceivedPayment` (see `payment` object above).

i'd like to know who paid with what crypto in my own database. i'll go with an api for now as you suggested.

lmk if i should open up an issue :)

**Comment by BamBam8619** [comment]
Hi I'm also looking for help in receiving my Bitcoin. Can anyone help me? I'm new here
